### PR TITLE
Add filters to os_router and os_subnet Fixes #37921 (#47220)

### LINKED ISSUE #37921: Multiple matches found even when using project parameter on openstack modules
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

os_router
os_subnet

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/racciari/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible
  executable location = /home/racciari/ansible2.5-test/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### CONFIGURATION

```
$ ansible-config dump --only-changed
DEFAULT_ROLES_PATH(/etc/ansible/ansible.cfg) = [u'/etc/ansible/roles', u'/usr/share/ansible/roles']
```

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When creating a router or a subnet with the os_router and os_subnet modules, it fails with `Multiple matches found for private_network` even when specifying the `project` parameter (that should filter the resource with the tenant_id/project_id) when multiple projects have the same router name/network name. It should not append when using the `project` parameter as this names are uniq in a project.

##### STEPS TO REPRODUCE

Using a clouds.yaml credentials file for Openstack and the following playbook. If you have multiple networks named `private_network` in different projects, it will fail.

```yaml
$ cat src/br.yml
- hosts: localhost
  tasks:
    - name: Create private subnet
      os_subnet:
        cloud: openio_openstack
        state: present
        network_name: private_network
        #network_name: 8323c509-61d5-4e3c-a3bd-38d03a97c1d5
        name: private_subnet
        project: ractest
        cidr: 10.0.0.0/8
        gateway_ip: 10.0.0.1
        allocation_pool_start: 10.0.0.2
        allocation_pool_end: 10.255.255.254
        dns_nameservers:
          - 8.8.4.4
          - 8.8.8.8
        wait: yes
```

##### EXPECTED RESULTS

When properly filtered, it works with an id (which is uniq) and with a name (even if not uniq in the domain).
```
$ ansible-playbook src/br.yml -vvvv
ansible-playbook 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/racciari/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible
  executable location = /home/racciari/ansible2.5-test/bin/ansible-playbook
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

Loading callback plugin default of type stdout, v2.0 from /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/plugins/callback/default.pyc

PLAYBOOK: br.yml ********************************************************************************************************************************************************************
1 plays in src/br.yml

PLAY [localhost] ********************************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************
task path: /home/racciari/ansible2.5-test/src/br.yml:1
Using module file /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: racciari
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001 `" && echo ansible-tmp-1522066063.63-271885951602001="` echo /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001 `" ) && sleep 0'
<127.0.0.1> PUT /home/racciari/.ansible/tmp/ansible-local-8597k8pjhO/tmpFPo0V9 TO /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001/ /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/racciari/ansible2.5-test/bin/python /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/racciari/.ansible/tmp/ansible-tmp-1522066063.63-271885951602001/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [Create private subnet] ********************************************************************************************************************************************************
task path: /home/racciari/ansible2.5-test/src/br.yml:3
Using module file /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/modules/cloud/openstack/os_subnet.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: racciari
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503 `" && echo ansible-tmp-1522066064.13-41849803861503="` echo /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503 `" ) && sleep 0'
<127.0.0.1> PUT /home/racciari/.ansible/tmp/ansible-local-8597k8pjhO/tmpBstinP TO /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503/os_subnet.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503/ /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503/os_subnet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/racciari/ansible2.5-test/bin/python /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503/os_subnet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/racciari/.ansible/tmp/ansible-tmp-1522066064.13-41849803861503/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false,
    "id": "789e75e5-22e9-4495-b229-c5788dcbd00d",
    "invocation": {
        "module_args": {
            "allocation_pool_end": "10.255.255.254",
            "allocation_pool_start": "10.0.0.2",
            "api_timeout": null,
            "auth": null,
            "auth_type": null,
            "availability_zone": null,
            "cacert": null,
            "cert": null,
            "cidr": "10.0.0.0/8",
            "dns_nameservers": [
                "8.8.4.4",
                "8.8.8.8"
            ],
            "enable_dhcp": true,
            "gateway_ip": "10.0.0.1",
            "host_routes": null,
            "interface": "public",
            "ip_version": "4",
            "ipv6_address_mode": null,
            "ipv6_ra_mode": null,
            "key": null,
            "name": "private_subnet",
            "network_name": "private_network",
            "no_gateway_ip": false,
            "region_name": null,
            "state": "present",
            "timeout": 180,
            "use_default_subnetpool": false,
            "verify": null,
            "wait": true
        }
    },
    "subnet": {
        "allocation_pools": [
            {
                "end": "10.255.255.254",
                "start": "10.0.0.2"
            }
        ],
        "cidr": "10.0.0.0/8",
        "created_at": "2018-03-22T20:52:20",
        "description": "",
        "dns_nameservers": [
            "8.8.4.4",
            "8.8.8.8"
        ],
        "enable_dhcp": true,
        "gateway_ip": "10.0.0.1",
        "host_routes": [],
        "id": "789e75e5-22e9-4495-b229-c5788dcbd00d",
        "ip_version": 4,
        "ipv6_address_mode": null,
        "ipv6_ra_mode": null,
        "name": "private_subnet",
        "network_id": "8323c509-61d5-4e3c-a3bd-38d03a97c1d5",
        "subnetpool_id": null,
        "tenant_id": "54c8b2e72d154490a20a8c9d20fcf33b",
        "updated_at": "2018-03-22T20:52:20"
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

##### ACTUAL RESULTS

```
$ ansible-playbook src/br.yml -vvvv
ansible-playbook 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/racciari/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible
  executable location = /home/racciari/ansible2.5-test/bin/ansible-playbook
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

Loading callback plugin default of type stdout, v2.0 from /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/plugins/callback/default.pyc

PLAYBOOK: br.yml ********************************************************************************************************************************************************************
1 plays in src/br.yml

PLAY [localhost] ********************************************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************************************
task path: /home/racciari/ansible2.5-test/src/br.yml:1
Using module file /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/modules/system/setup.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: racciari
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187 `" && echo ansible-tmp-1522065739.82-211048661628187="` echo /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187 `" ) && sleep 0'
<127.0.0.1> PUT /home/racciari/.ansible/tmp/ansible-local-8511n0fK3C/tmpzIAe0Z TO /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187/setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187/ /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/racciari/ansible2.5-test/bin/python /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187/setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/racciari/.ansible/tmp/ansible-tmp-1522065739.82-211048661628187/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [Create private subnet] ********************************************************************************************************************************************************
task path: /home/racciari/ansible2.5-test/src/br.yml:3
Using module file /home/racciari/ansible2.5-test/lib/python2.7/site-packages/ansible/modules/cloud/openstack/os_subnet.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: racciari
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206 `" && echo ansible-tmp-1522065740.31-102330667281206="` echo /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206 `" ) && sleep 0'
<127.0.0.1> PUT /home/racciari/.ansible/tmp/ansible-local-8511n0fK3C/tmpD2T21R TO /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206/os_subnet.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206/ /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206/os_subnet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/racciari/ansible2.5-test/bin/python /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206/os_subnet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/racciari/.ansible/tmp/ansible-tmp-1522065740.31-102330667281206/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_LOVl1r/ansible_module_os_subnet.py", line 341, in main
    if _needs_update(subnet, module, cloud):
  File "/tmp/ansible_LOVl1r/ansible_module_os_subnet.py", line 193, in _needs_update
    _can_update(subnet, module, cloud)
  File "/tmp/ansible_LOVl1r/ansible_module_os_subnet.py", line 172, in _can_update
    network = cloud.get_network(network_name)

▽
  File "/home/racciari/ansible2.5-test/lib/python2.7/site-packages/shade/openstackcloud.py", line 2750, in get_network
    return _utils._get_entity(self, 'network', name_or_id, filters)
  File "/home/racciari/ansible2.5-test/lib/python2.7/site-packages/shade/_utils.py", line 241, in _get_entity
    "Multiple matches found for %s" % name_or_id)

fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allocation_pool_end": "10.255.255.254",
            "allocation_pool_start": "10.0.0.2",
            "api_timeout": null,
            "auth": null,
            "auth_type": null,
            "availability_zone": null,
            "cacert": null,
            "cert": null,
            "cidr": "10.0.0.0/8",
            "dns_nameservers": [
                "8.8.4.4",
                "8.8.8.8"
            ],
            "enable_dhcp": true,
            "gateway_ip": "10.0.0.1",
            "host_routes": null,
            "interface": "public",
            "ip_version": "4",
            "ipv6_address_mode": null,
            "ipv6_ra_mode": null,
            "key": null,
            "name": "private_subnet",
            "network_name": "private_network",
            "no_gateway_ip": false,
            "region_name": null,
            "state": "present",
            "timeout": 180,
            "use_default_subnetpool": false,
            "verify": null,
            "wait": true
        }
    },
    "msg": "Multiple matches found for private_network"
}
	to retry, use: --limit @/home/racciari/ansible2.5-test/src/br.retry

PLAY RECAP **************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

### PR DESCRIPTION
**Pull request by GeorginaShippey**
Would like this bugfix backported into 2.7

### CODE REVIEW

**Comment by abadger** [comment]
Informational: backport of https://github.com/ansible/ansible/pull/47220  (which is already in 2.8.x as it was merged before stable-2.8 branched).

@GeorginaShippey  2.7.x is only open for critical bugfixes and security fixes.  At first glance, this doesn't look like that's the case here.  Is there something that makes it critical?  (Like it affects core code or is an issue that is affecting a ton of users?)  If it's not critical, we'd usually advise users to update to the newer version, 2.8.x in order to fix the problem.

**Comment by jrosser** [comment]
@abadger I appreciate that 2.7.x is closed - however we are in a rather unfortunate position here because the patch has been backported to 2.6.x (and possibly earlier), exists in 2.8.x but was never merged to 2.7.x.

Life is never quite as simple as just updating to 2.8.x, some notably major ansible projects such as ceph-ansible and openstack-ansible pin their ansible version and own major release versions together and do not modify released stable branches to accommodate new ansible versions.

Whilst it may be arguable that this is neither a critical bugfix nor a security fix, it does not seem right to leave a regression in 2.7.x when earlier and later branches include the fix. My particular interest is that this bug affects users of the currently released Stein version of openstack-ansible who are pinned to ansible 2.7.x for the lifetime of that release.

**Comment by abadger** [comment]
@GeorginaShippey or @jrosser Could you take this to one of the IRC meetings to ask if this should qualify as a critical bug (critical bugs trigger of enough importance that they trigger a new 2.7.x release)?  A regression within the 2.7.x line would be a good case.  A regression from 2.6.x is not as strong but sounds like it might be the best basis you have.  The counter argument to that is that 2.7.x has survived with this bug in place for ten months without a backport being mentioned so that depresses the chances that this would be considered a critical bug.  (As does the fact that this isn't a bug in the core engine).  At the meeting, the core team can take a vote on whether they'd like to consider this a critical bug to trigger a release.

Meetings are Tuesdays and Thursdays: https://github.com/ansible/community/tree/master/meetings#tuesdays

File: lib/ansible/modules/cloud/openstack/os_router.py [hunk 1]
```diff
@@ -414,10 +414,10 @@ def main():
 
         # Validate and cache the subnet IDs so we can avoid duplicate checks
         # and expensive API calls.
-        external_ids, subnet_internal_ids, internal_portids = _validate_subnets(module, cloud)
+        external_ids, subnet_internal_ids, internal_portids = _validate_subnets(module, cloud, filters)
         if module.check_mode:
             module.exit_json(
-                changed=_system_state_change(cloud, module, router, net, subnet_internal_ids, internal_portids)
+                changed=_system_state_change(cloud, module, router, net, subnet_internal_ids, internal_portids, filters)
```
Comment (cloudnull) [inline lib/ansible/modules/cloud/openstack/os_router.py:420]: nit: I think this should be wrapped. 121 line length is longer than it should be. While Ansible has a lot of code which doesn't adhere to pep8 we should strive to be better, so lets try and keep the line length to 80 or 120.

File: lib/ansible/modules/cloud/openstack/os_router.py [hunk 2]
```diff
@@ -436,7 +436,7 @@ def main():
                     cloud.add_router_interface(router, port_id=int_p_id)
                 changed = True
             else:
-                if _needs_update(cloud, module, router, net, subnet_internal_ids, internal_portids):
+                if _needs_update(cloud, module, router, net, subnet_internal_ids, internal_portids, filters):
```
Comment (cloudnull) [inline lib/ansible/modules/cloud/openstack/os_router.py:439]: nit: I think this should be wrapped. 121 line length is longer than it should be. While Ansible has a lot of code which doesn't adhere to pep8 we should strive to be better, so lets try and keep the line length to 80 or 120.

File: lib/ansible/modules/cloud/openstack/os_router.py [hunk 3]
```diff
@@ -453,7 +453,7 @@ def main():
                         for port in ports:
                             cloud.remove_router_interface(router, port_id=port['id'])
                     if internal_portids:
-                        external_ids, subnet_internal_ids, internal_portids = _validate_subnets(module, cloud)
+                        external_ids, subnet_internal_ids, internal_portids = _validate_subnets(module, cloud, filters)
```
Comment (cloudnull) [inline lib/ansible/modules/cloud/openstack/os_router.py:456]: nit: I think this should be wrapped. 121 line length is longer than it should be. While Ansible has a lot of code which doesn't adhere to pep8 we should strive to be better, so lets try and keep the line length to 80 or 120.

**Review by cloudnull — CHANGES_REQUESTED** [review]
couple of nits but otherwise looks good to me.

**Comment by cloudnull** [comment]
@abadger was this change meant to be closed?

**Pull request closed without merging**